File: src/main/java/ru/n5g/learningenglish/words/IrregularVerbsWords.java
package ru.n5g.learningenglish.words;

import java.util.*;

/**
 * Неправильные глаголы
 *
 * @author Gleb Belyaev
 */
public class IrregularVerbsWords extends WordsAbs<String, String[]> {

    protected Map<String, String[]> originalMap;

    public IrregularVerbsWords() {
        super();
        originalMap = new LinkedHashMap<String, String[]>(map);
    }

    protected Map<String, String[]> initializationWords() {
        Map<String, String[]> words = new LinkedHashMap<String, String[]>();
        words.put("быть", new String[]{"be", "was were", "been"});
        words.put("бить", new String[]{"beat", "beat", "beaten"});
        words.put("становиться", new String[]{"become", "became", "become"});
        words.put("начинать", new String[]{"begin", "began", "begun"});
        words.put("кровоточить", new String[]{"bleed", "bled", "bled"});
        words.put("дуть", new String[]{"blow", "blew", "blown"});
        words.put("ломать", new String[]{"break", "broke", "broken"});
        words.put("приносить", new String[]{"bring", "brought", "brought"});
        words.put("строить", new String[]{"build", "built", "built"});
        words.put("гореть", new String[]{"burn", "burnt", "burnt"});
        words.put("ворваться, разрываться, взрываться", new String[]{"burst", "burst", "burst"});
        words.put("покупать", new String[]{"buy", "bought", "bought"});
        words.put("предлагать цену", new String[]{"bid", "bid", "bid"});
        words.put("кусать", new String[]{"bite", "bit", "bitten"});
        words.put("ловить, хватать, успеть", new String[]{"catch", "caught", "caught"});
        words.put("выбирать", new String[]{"choose", "chose", "chosen"});
        words.put("приходить", new String[]{"come", "came", "come"});
        words.put("стоить", new String[]{"cost", "cost", "cost"});
        words.put("ползать", new String[]{"creep", "crept", "crept"});
        words.put("резать", new String[]{"cut", "cut", "cut"});
        words.put("делать", new String[]{"do", "did", "done"});
        words.put("рисовать, тащить", new String[]{"draw", "drew", "drawn"});
        words.put("мечтать, дремать", new String[]{"dream", "dreamt", "dreamt"});
        words.put("пить", new String[]{"drink", "drank", "drunk"});
        words.put("водить", new String[]{"drive", "drove", "driven"});
        words.put("есть", new String[]{"eat", "ate", "eaten"});
        words.put("падать", new String[]{"fall", "fell", "fallen"});
        words.put("кормить", new String[]{"feed", "fed", "fed"});
        words.put("чувствовать", new String[]{"feel", "felt", "felt"});
        words.put("бороться", new String[]{"fight", "fought", "fought"});
        words.put("находить", new String[]{"find", "found", "found"});
        words.put("соответствовать, подходить", new String[]{"fit", "fit", "fit"});
        words.put("летать", new String[]{"fly", "flew", "flown"});
        words.put("забывать", new String[]{"forget", "forgot", "forgotten"});
        words.put("прощать", new String[]{"forgive", "forgave", "forgiven"});
        words.put("замерзать", new String[]{"freeze", "froze", "frozen"});
        words.put("получать", new String[]{"get", "got", "got"});
        words.put("давать", new String[]{"give", "gave", "given"});
        words.put("идти", new String[]{"go", "went", "gone"});
        words.put("расти", new String[]{"grow", "grew", "grown"});
        words.put("вешать", new String[]{"hang", "hung", "hung"});
        words.put("иметь", new String[]{"have", "had", "had"});
        words.put("слышать", new String[]{"hear", "heard", "heard"});
        words.put("прятать", new String[]{"hide", "hid", "hidden"});
        words.put("попадать в цель", new String[]{"hit", "hit", "hit"});
        words.put("держать", new String[]{"hold", "held", "held"});
        words.put("ушибить", new String[]{"hurt", "hurt", "hurt"});
        words.put("содержать", new String[]{"keep", "keep", "keep"});
        words.put("стоять на коленях", new String[]{"kneel", "knelt", "knelt"});
        words.put("знать", new String[]{"know", "knew", "known"});
        words.put("класть", new String[]{"lay", "laid", "laid"});
        words.put("вести", new String[]{"lead", "led", "led"});
        words.put("наклоняться, опираться, прислоняться", new String[]{"lean", "leant", "leant"});
        words.put("учить", new String[]{"learn", "learnt", "learnt"});
        words.put("оставлять", new String[]{"leave", "left", "left"});
        words.put("занимать", new String[]{"lend", "lent", "lent"});
        words.put("позволять", new String[]{"let", "let", "let"});
        words.put("лежать", new String[]{"lie", "lay", "lain"});
        words.put("освещать", new String[]{"light", "lit", "lit"});
        words.put("терять", new String[]{"lose", "lost", "lost"});
        words.put("производить", new String[]{"make", "made", "made"});
        words.put("значить", new String[]{"mean", "meant", "meant"});
        words.put("встречать", new String[]{"meet", "met", "met"});
        words.put("ошибаться", new String[]{"mistake", "mistook", "mistaken"});
        words.put("платить", new String[]{"pay", "paid", "paid"});
        words.put("доказывать", new String[]{"prove", "proved", "proven"});
        words.put("положить", new String[]{"put", "put", "put"});
        words.put("выходить", new String[]{"quit", "quit", "quit"});
        words.put("читать", new String[]{"read", "read", "read"});
        words.put("ездить верхом", new String[]{"ride", "rode", "ridden"});
        words.put("звенеть", new String[]{"ring", "rang", "rung"});
        words.put("подниматься", new String[]{"rise", "rose", "risen"});
        words.put("бежать", new String[]{"run", "ran", "run"});
        words.put("сказать, говорить, утверждать", new String[]{"say", "said", "said"});
        words.put("видеть", new String[]{"see", "saw", "seen"});
        words.put("искать", new String[]{"seek", "sought", "sought"});
        words.put("продавать", new String[]{"sell", "sold", "sold"});
        words.put("посылать", new String[]{"send", "sent", "sent"});
        words.put("ставить", new String[]{"set", "set", "set"});
        words.put("шить", new String[]{"sew", "sewed", "sewn"});
        words.put("встряхивать", new String[]{"shake", "shook", "shaken"});
        words.put("показывать", new String[]{"show", "showed", "shown"});
        words.put("уменьшать", new String[]{"shrink", "shrank", "shrunk"});
        words.put("закрывать", new String[]{"shut", "shut", "shut"});
        words.put("петь", new String[]{"sing", "sang", "sung"});
        words.put("тонуть", new String[]{"sink", "sank sunk", "sunk"});
        words.put("сидеть", new String[]{"sit", "sat", "sat"});
        words.put("спать", new String[]{"sleep", "slept", "slept"});
        words.put("скользить", new String[]{"slide", "slid", "slid"});
        words.put("сеять", new String[]{"sow", "sowed", "sown"});
        words.put("говорить, выступать, разговаривать", new String[]{"speak", "spoke", "spoken"});
        words.put("произносить по буквам", new String[]{"spell", "spelt", "spelt"});
        words.put("тратить", new String[]{"spend", "spent", "spent"});
        words.put("проливать", new String[]{"spill", "spilt", "spilt"});
        words.put("портить", new String[]{"spoil", "spoilt", "spoilt"});
        words.put("расстилать", new String[]{"spread", "spread", "spread"});
        words.put("прыгать, вскочить, возникнуть", new String[]{"spring", "sprang", "sprung"});
        words.put("стоять", new String[]{"stand", "stood", "stood"});
        words.put("красть", new String[]{"steal", "stole", "stolen"});
        words.put("колоть", new String[]{"stick", "stuck", "stuck"});
        words.put("жалить", new String[]{"sting", "stung", "stung"});
        words.put("выметать", new String[]{"sweep", "swept", "swept"});
        words.put("разбухать", new String[]{"swell", "swelled", "swollen"});
        words.put("плавать", new String[]{"swim", "swam", "swum"});
        words.put("качать", new String[]{"swing", "swung", "swung"});
        words.put("брать, взять", new String[]{"take", "took", "taken"});
        words.put("учить, обучать", new String[]{"teach", "taught", "taught"});
        words.put("рвать", new String[]{"tear", "tore", "torn"});
        words.put("рассказывать", new String[]{"tell", "told", "told"});
        words.put("думать", new String[]{"think", "thought", "thought"});
        words.put("бросать", new String[]{"throw", "threw", "thrown"});
        words.put("понимать", new String[]{"understand", "understood", "understood"});
        words.put("предпринимать", new String[]{"undertake", "undertook", "undertaken"});
        words.put("просыпаться", new String[]{"wake", "woke", "woken"});
        words.put("носить", new String[]{"wear", "wore", "worn"});
        words.put("плакать", new String[]{"weep", "wept", "wept"});
        words.put("мочить", new String[]{"wet", "wet", "wet"});
        words.put("выигрывать", new String[]{"win", "won", "won"});
        words.put("извиваться", new String[]{"wind", "wound", "wound"});
        words.put("писать", new String[]{"write", "wrote", "written"});
        return words;
    }

    public void setIntervalWords(int startWord, int endWord) {
        map = new LinkedHashMap<String, String[]>();
        List<String> keys = new ArrayList<String>(originalMap.keySet());
        for (int i = 0; i < keys.size(); i++) {
            if (i >= startWord && i < endWord) {
                map.put(keys.get(i), originalMap.get(keys.get(i)));
            }
        }
    }

    @Override
    public boolean isSound() {
        return true;
    }

    @Override
    public String pathSound() {
        return "irregularverbs";
    }
}

File: src/main/java/ru/n5g/learningenglish/words/IrregularVerbsWords.kt
package ru.n5g.learningenglish.words

import java.util.LinkedHashMap


data class IrregularVerbWord(val index: Int, //позиция в словоре
                             val translate: String, //перевод
                             val infinitive: String, //неопределенная форма глагола
                             val pastSimple: String, //простое прошедшее время
                             val pastParticiple: String) //простое прошедшее время

class IrregularVerbsWords{
    val words = arrayListOf<IrregularVerbWord>()
    init {
        var indx = 0
        for ((t, v) in words()) {
            words.add(IrregularVerbWord(indx++, t, v[0], v[1], v[2]))
        }
    }

    fun getWords(indxStart: Int, indxEnd: Int): List<IrregularVerbWord> {
        return words.asSequence().filter { it.index in indxStart..indxEnd }.toList()
    }

    fun count() = words.size
}

fun words(): LinkedHashMap<String, Array<String>> {
    val words = LinkedHashMap<String, Array<String>>()
    words["быть"] = arrayOf("be", "was were", "been")
    words["бить"] = arrayOf("beat", "beat", "beaten")
    words["становиться"] = arrayOf("become", "became", "become")
    words["начинать"] = arrayOf("begin", "began", "begun")
    words["кровоточить"] = arrayOf("bleed", "bled", "bled")
    words["дуть"] = arrayOf("blow", "blew", "blown")
    words["ломать"] = arrayOf("break", "broke", "broken")
    words["приносить"] = arrayOf("bring", "brought", "brought")
    words["строить"] = arrayOf("build", "built", "built")
    words["гореть"] = arrayOf("burn", "burnt", "burnt")
    words["ворваться, разрываться, взрываться"] = arrayOf("burst", "burst", "burst")
    words["покупать"] = arrayOf("buy", "bought", "bought")
    words["предлагать цену"] = arrayOf("bid", "bid", "bid")
    words["кусать"] = arrayOf("bite", "bit", "bitten")
    words["ловить, хватать, успеть"] = arrayOf("catch", "caught", "caught")
    words["выбирать"] = arrayOf("choose", "chose", "chosen")
    words["приходить"] = arrayOf("come", "came", "come")
    words["стоить"] = arrayOf("cost", "cost", "cost")
    words["ползать"] = arrayOf("creep", "crept", "crept")
    words["резать"] = arrayOf("cut", "cut", "cut")
    words["делать"] = arrayOf("do", "did", "done")
    words["рисовать, тащить"] = arrayOf("draw", "drew", "drawn")
    words["мечтать, дремать"] = arrayOf("dream", "dreamt", "dreamt")
    words["пить"] = arrayOf("drink", "drank", "drunk")
    words["водить"] = arrayOf("drive", "drove", "driven")
    words["есть"] = arrayOf("eat", "ate", "eaten")
    words["падать"] = arrayOf("fall", "fell", "fallen")
    words["кормить"] = arrayOf("feed", "fed", "fed")
    words["чувствовать"] = arrayOf("feel", "felt", "felt")
    words["бороться"] = arrayOf("fight", "fought", "fought")
    words["находить"] = arrayOf("find", "found", "found")
    words["соответствовать, подходить"] = arrayOf("fit", "fit", "fit")
    words["летать"] = arrayOf("fly", "flew", "flown")
    words["забывать"] = arrayOf("forget", "forgot", "forgotten")
    words["прощать"] = arrayOf("forgive", "forgave", "forgiven")
    words["замерзать"] = arrayOf("freeze", "froze", "frozen")
    words["получать"] = arrayOf("get", "got", "got")
    words["давать"] = arrayOf("give", "gave", "given")
    words["идти"] = arrayOf("go", "went", "gone")
    words["расти"] = arrayOf("grow", "grew", "grown")
    words["вешать"] = arrayOf("hang", "hung", "hung")
    words["иметь"] = arrayOf("have", "had", "had")
    words["слышать"] = arrayOf("hear", "heard", "heard")
    words["прятать"] = arrayOf("hide", "hid", "hidden")
    words["попадать в цель"] = arrayOf("hit", "hit", "hit")
    words["держать"] = arrayOf("hold", "held", "held")
    words["ушибить"] = arrayOf("hurt", "hurt", "hurt")
    words["держать, хранить, беречь, иметь, содержать"] = arrayOf("keep", "kept", "kept")
    words["становиться на колени"] = arrayOf("kneel", "knelt", "knelt")
    words["знать"] = arrayOf("know", "knew", "known")
    words["класть, положить, покрывать"] = arrayOf("lay", "laid", "laid")
    words["вести за собой, сопровождать, руководить"] = arrayOf("lead", "led", "led")
    words["наклоняться, опираться, прислоняться"] = arrayOf("lean", "leant", "leant")
    words["учить"] = arrayOf("learn", "learnt", "learnt")
    words["оставлять"] = arrayOf("leave", "left", "left")
    words["одалживать, давать взаймы (в долг)"] = arrayOf("lend", "lent", "lent")
    words["позволять, разрешать"] = arrayOf("let", "let", "let")
    words["лежать"] = arrayOf("lie", "lay", "lain")
    words["освещать"] = arrayOf("light", "lit", "lit")
    words["терять, лишаться, утрачивать"] = arrayOf("lose", "lost", "lost")
    words["производить"] = arrayOf("make", "made", "made")
    words["значить"] = arrayOf("mean", "meant", "meant")
    words["встречать"] = arrayOf("meet", "met", "met")
    words["ошибаться"] = arrayOf("mistake", "mistook", "mistaken")
    words["платить"] = arrayOf("pay", "paid", "paid")
    words["доказывать, оказаться"] = arrayOf("prove", "proved", "proven")
    words["положить"] = arrayOf("put", "put", "put")
    words["выходить"] = arrayOf("quit", "quit", "quit")
    words["читать"] = arrayOf("read", "read", "read")
    words["ездить верхом"] = arrayOf("ride", "rode", "ridden")
    words["звенеть"] = arrayOf("ring", "rang", "rung")
    words["подниматься"] = arrayOf("rise", "rose", "risen")
    words["бежать"] = arrayOf("run", "ran", "run")
    words["сказать, говорить, утверждать"] = arrayOf("say", "said", "said")
    words["видеть"] = arrayOf("see", "saw", "seen")
    words["искать"] = arrayOf("seek", "sought", "sought")
    words["продавать"] = arrayOf("sell", "sold", "sold")
    words["посылать"] = arrayOf("send", "sent", "sent")
    words["ставить"] = arrayOf("set", "set", "set")
    words["шить"] = arrayOf("sew", "sewed", "sewn")
    words["встряхивать"] = arrayOf("shake", "shook", "shaken")
    words["показывать"] = arrayOf("show", "showed", "shown")
    words["уменьшать"] = arrayOf("shrink", "shrank", "shrunk")
    words["закрывать"] = arrayOf("shut", "shut", "shut")
    words["петь"] = arrayOf("sing", "sang", "sung")
    words["тонуть"] = arrayOf("sink", "sank sunk", "sunk")
    words["сидеть"] = arrayOf("sit", "sat", "sat")
    words["спать"] = arrayOf("sleep", "slept", "slept")
    words["скользить"] = arrayOf("slide", "slid", "slid")
    words["сеять"] = arrayOf("sow", "sowed", "sown")
    words["говорить, выступать, разговаривать"] = arrayOf("speak", "spoke", "spoken")
    words["произносить по буквам"] = arrayOf("spell", "spelt", "spelt")
    words["тратить"] = arrayOf("spend", "spent", "spent")
    words["проливать"] = arrayOf("spill", "spilt", "spilt")
    words["портить"] = arrayOf("spoil", "spoilt", "spoilt")
    words["расстилать"] = arrayOf("spread", "spread", "spread")
    words["прыгать, вскочить, возникнуть"] = arrayOf("spring", "sprang", "sprung")
    words["стоять"] = arrayOf("stand", "stood", "stood")
    words["красть"] = arrayOf("steal", "stole", "stolen")
    words["колоть"] = arrayOf("stick", "stuck", "stuck")
    words["жалить"] = arrayOf("sting", "stung", "stung")
    words["выметать"] = arrayOf("sweep", "swept", "swept")
    words["разбухать"] = arrayOf("swell", "swelled", "swollen")
    words["плавать"] = arrayOf("swim", "swam", "swum")
    words["качать"] = arrayOf("swing", "swung", "swung")
    words["брать, взять"] = arrayOf("take", "took", "taken")
    words["учить, обучать"] = arrayOf("teach", "taught", "taught")
    words["рвать"] = arrayOf("tear", "tore", "torn")
    words["рассказывать"] = arrayOf("tell", "told", "told")
    words["думать"] = arrayOf("think", "thought", "thought")
    words["бросать"] = arrayOf("throw", "threw", "thrown")
    words["понимать"] = arrayOf("understand", "understood", "understood")
    words["предпринимать"] = arrayOf("undertake", "undertook", "undertaken")
    words["просыпаться"] = arrayOf("wake", "woke", "woken")
    words["носить"] = arrayOf("wear", "wore", "worn")
    words["плакать"] = arrayOf("weep", "wept", "wept")
    words["мочить"] = arrayOf("wet", "wet", "wet")
    words["выигрывать"] = arrayOf("win", "won", "won")
    words["извиваться"] = arrayOf("wind", "wound", "wound")
    words["писать"] = arrayOf("write", "wrote", "written")
    return words
}
